File: sdk/typescript/worker.mjs
import { pathToFileURL } from "node:url";
import { encode, decode } from "./cbor.mjs";
import { SurfaceBuilder, RefractionBeam, OpticalContext, err, ok } from "./index.mjs";

const entryIndex = process.argv.indexOf("--entry");
if (entryIndex < 0 || !process.argv[entryIndex + 1]) throw new Error("--entry is required");
const module = await import(pathToFileURL(process.argv[entryIndex + 1]).href);
const lens = module.default;
if (!lens?.id || typeof lens.view !== "function" || typeof lens.refract !== "function")
  throw new Error("entry must default-export a Lens with id, view and refract");

const controllers = new Map();
const write = (frame) => {
  const payload = encode(frame); const header = Buffer.alloc(4); header.writeUInt32BE(payload.length);
  process.stdout.write(Buffer.concat([header, payload]));
};
const failure = (error) => ({ ok: false, error: {
  code: error?.code ?? "lens_crashed", message: String(error?.message ?? error), retryable: false,
}});

let pending = Buffer.alloc(0);
const chunks = process.stdin[Symbol.asyncIterator]();
const readFrame = async () => {
  while (true) {
    if (pending.length >= 4) {
      const size = pending.readUInt32BE(0);
      if (size <= 0 || size > 16 * 1024 * 1024) throw new Error("invalid frame size");
      if (pending.length >= 4 + size) {
        const frame = decode(pending.subarray(4, 4 + size));
        pending = pending.subarray(4 + size);
        return frame;
      }
    }
    const next = await chunks.next();
    if (next.done) return null;
    pending = Buffer.concat([pending, next.value]);
  }
};

while (true) {
    const frame = await readFrame();
    if (!frame) break;
    const response = { request_id: frame.request_id, payload: {} };
    try {
      if (frame.type === "worker.hello") {
        response.type = "worker.ready";
        response.payload = { protocol_major: 1, protocol_minor: 1, lens_id: lens.id,
          runtime: "node", runtime_version: process.versions.node,
          version: lens.version ?? "0.1.0",
          features: { derive: typeof lens.derive === "function",
            coordinate: typeof lens.coordinate === "function",
            query: typeof lens.query === "function" } };
      } else if (frame.type === "lens.view.request") {
        const surface = new SurfaceBuilder(lens.id);
        const result = await lens.view(frame.payload.window, surface);
        response.type = "lens.view.result";
        response.payload = result?.ok === false ? result : { ok: true,
          surface: { epoch: frame.payload.epoch ?? 0,
            contributions: surface.contributions, proposals: surface.proposals } };
      } else if (frame.type === "lens.derive.request") {
        const result = typeof lens.derive === "function" ?
          await lens.derive(frame.payload.window) : ok(null);
        response.type = "lens.derive.result";
        response.payload = result?.ok === false ? result :
          { ok: true, state: result && typeof result === "object" && "ok" in result ?
              result.value : result ?? null };
      } else if (frame.type === "lens.coordinate.request") {
        const exchange = async (payload) => {
          write({ type: "host.optical.request", request_id: frame.request_id, payload });
          const hostResponse = await readFrame();
          if (!hostResponse || hostResponse.type !== "host.optical.result")
            return err("protocol_error", "missing optical host response");
          if (!hostResponse.payload?.ok) {
            const failure = hostResponse.payload?.error ?? {};
            return err(failure.code ?? "provider_failed",
              failure.message ?? "optical host failed", failure.retryable ?? false);
          }
          return ok(hostResponse.payload.value);
        };
        const surface = new SurfaceBuilder(lens.id);
        const result = typeof lens.coordinate === "function" ?
          await lens.coordinate(frame.payload.window, new OpticalContext(exchange), surface) :
          ok(undefined);
        response.type = "lens.coordinate.result";
        response.payload = result?.ok === false ? result : { ok: true,
          surface: { epoch: 0, contributions: surface.contributions,
            proposals: surface.proposals } };
      } else if (frame.type === "lens.query.request") {
        const result = typeof lens.query === "function" ? await lens.query(
          frame.payload.state?.value, frame.payload.capability,
          frame.payload.parameters, frame.payload.budget) :
          err("unsupported", `unsupported optical capability: ${frame.payload.capability}`);
        response.type = "lens.query.result";
        response.payload = result?.ok === false ? result :
          { ok: true, value: result && typeof result === "object" && "ok" in result ?
              result.value : result };
      } else if (frame.type === "lens.refract.request") {
        const controller = new AbortController(); controllers.set(frame.request_id, controller);
        const beam = new RefractionBeam(frame.payload.act, controller.signal);
        const result = await lens.refract(frame.payload.window, frame.payload.act, beam);
        controllers.delete(frame.request_id); response.type = "lens.refract.result";
        response.payload = result?.ok === false ? result : { ok: true,
          ...(result?.value ?? result), drafts: beam.drafts, logs: beam.logs };
      } else if (frame.type === "beam.cancel") {
        controllers.get(frame.payload.request_id)?.abort(); response.type = "beam.cancelled";
        response.payload = { ok: true };
      } else if (frame.type === "worker.shutdown") {
        for (const controller of controllers.values()) controller.abort();
        response.type = "worker.stopped"; response.payload = { ok: true }; write(response); process.exit(0);
      } else throw new Error(`unknown frame type: ${frame.type}`);
    } catch (error) {
      response.type = ({
        "lens.view.request": "lens.view.result",
        "lens.derive.request": "lens.derive.result",
        "lens.coordinate.request": "lens.coordinate.result",
        "lens.query.request": "lens.query.result",
      })[frame.type] ?? "lens.refract.result";
      response.payload = failure(error);
    }
    write(response);
}

File: sdk/python/tokmon_lens_sdk/worker.py
from __future__ import annotations

import argparse
import asyncio
import importlib.util
import inspect
import struct
import sys
from pathlib import Path
from typing import Any

if __package__ in (None, ""):
    sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from tokmon_lens_sdk.cbor import decode, encode
from tokmon_lens_sdk.core import Lens, OpticalContext, RefractionBeam, Result, SurfaceBuilder, err, ok


def load_lens(path: Path) -> Any:
    spec = importlib.util.spec_from_file_location("tokmon_user_lens", path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"cannot load Lens entry: {path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    lens = getattr(module, "lens", None)
    if lens is None:
        lens_type = getattr(module, "LensEntry", None)
        lens = lens_type() if lens_type else None
    if lens is None or not getattr(lens, "id", None):
        raise RuntimeError("entry must export 'lens' or LensEntry")
    return lens


def result_payload(result: Any) -> dict[str, Any]:
    if isinstance(result, Result):
        if not result.ok:
            assert result.error is not None
            return {"ok": False, "error": result.error.__dict__}
        value = result.value
    else:
        value = result
    if isinstance(value, dict):
        return {"ok": True, **value}
    return {"ok": True, "value": value}


def optical_result_payload(result: Any, field: str) -> dict[str, Any]:
    if isinstance(result, Result):
        if not result.ok:
            assert result.error is not None
            return {"ok": False, "error": result.error.__dict__}
        result = result.value
    return {"ok": True, field: result}


async def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--entry", type=Path, required=True)
    args = parser.parse_args()
    lens = load_lens(args.entry.resolve())
    tasks: dict[int, tuple[asyncio.Task[Any], RefractionBeam]] = {}
    reader = sys.stdin.buffer
    writer = sys.stdout.buffer

    async def read_exact(size: int) -> bytes:
        chunks: list[bytes] = []
        remaining = size
        while remaining:
            chunk = await asyncio.to_thread(reader.read, remaining)
            if not chunk:
                break
            chunks.append(chunk)
            remaining -= len(chunk)
        return b"".join(chunks)

    async def read_worker_frame() -> dict[str, Any] | None:
        header = await read_exact(4)
        if not header:
            return None
        if len(header) != 4:
            raise RuntimeError("truncated worker header")
        size = struct.unpack(">I", header)[0]
        if size <= 0 or size > 16 * 1024 * 1024:
            raise RuntimeError("invalid worker frame size")
        payload = await read_exact(size)
        if len(payload) != size:
            raise RuntimeError("truncated worker frame")
        return decode(payload)

    def write_worker_frame(frame: dict[str, Any]) -> None:
        packed = encode(frame)
        writer.write(struct.pack(">I", len(packed)) + packed)
        writer.flush()

    while True:
        frame = await read_worker_frame()
        if frame is None:
            return 0
        request_id = int(frame.get("request_id", 0))
        response: dict[str, Any] = {"request_id": request_id, "payload": {}}
        try:
            if frame["type"] == "worker.hello":
                def overrides(name: str) -> bool:
                    method = getattr(lens, name, None)
                    if not callable(method):
                        return False
                    if not isinstance(lens, Lens):
                        return True
                    return getattr(type(lens), name, None) is not getattr(Lens, name, None)

                response.update(type="worker.ready", payload={
                    "protocol_major": 1, "protocol_minor": 1, "lens_id": lens.id,
                    "runtime": "cpython",
                    "runtime_version": ".".join(str(value) for value in sys.version_info[:3]),
                    "version": getattr(lens, "version", "0.1.0"),
                    "features": {"derive": overrides("derive"),
                                 "coordinate": overrides("coordinate"),
                                 "query": overrides("query")},
                })
            elif frame["type"] == "lens.view.request":
                surface = SurfaceBuilder(lens.id)
                result = lens.view(frame["payload"]["window"], surface)
                response.update(type="lens.view.result", payload=result_payload(result))
                if response["payload"].get("ok"):
                    response["payload"]["surface"] = {
                        "epoch": frame["payload"].get("epoch", 0),
                        "contributions": surface.contributions,
                        "proposals": surface.proposals,
                    }
            elif frame["type"] == "lens.derive.request":
                result = lens.derive(frame["payload"]["window"])
                if inspect.isawaitable(result):
                    result = await result
                response.update(type="lens.derive.result",
                                payload=optical_result_payload(result, "state"))
            elif frame["type"] == "lens.coordinate.request":
                async def exchange(payload: dict[str, Any]) -> Result[Any]:
                    write_worker_frame({"type": "host.optical.request",
                                        "request_id": request_id, "payload": payload})
                    host_response = await read_worker_frame()
                    if host_response is None or host_response.get("type") != "host.optical.result":
                        return err("protocol_error", "missing optical host response")
                    host_payload = host_response.get("payload", {})
                    if not host_payload.get("ok"):
                        failure = host_payload.get("error", {})
                        return err(str(failure.get("code", "provider_failed")),
                                   str(failure.get("message", "optical host failed")),
                                   bool(failure.get("retryable", False)))
                    return ok(host_payload.get("value"))
                surface = SurfaceBuilder(lens.id)
                result = lens.coordinate(frame["payload"]["window"],
                                         OpticalContext(exchange), surface)
                if inspect.isawaitable(result):
                    result = await result
                response.update(type="lens.coordinate.result", payload=result_payload(result))
                if response["payload"].get("ok"):
                    response["payload"]["surface"] = {
                        "epoch": 0, "contributions": surface.contributions,
                        "proposals": surface.proposals,
                    }
            elif frame["type"] == "lens.query.request":
                result = lens.query(frame["payload"]["state"].get("value"),
                                    frame["payload"]["capability"],
                                    frame["payload"]["parameters"],
                                    frame["payload"]["budget"])
                if inspect.isawaitable(result):
                    result = await result
                response.update(type="lens.query.result",
                                payload=optical_result_payload(result, "value"))
            elif frame["type"] == "lens.refract.request":
                beam = RefractionBeam(frame["payload"]["act"])
                task = asyncio.create_task(lens.refract(
                    frame["payload"]["window"], frame["payload"]["act"], beam
                ))
                tasks[request_id] = (task, beam)
                result = await task
                tasks.pop(request_id, None)
                response.update(type="lens.refract.result", payload=result_payload(result))
                if response["payload"].get("ok"):
                    response["payload"].update(drafts=beam.drafts, logs=beam.logs)
            elif frame["type"] == "beam.cancel":
                target = int(frame["payload"].get("request_id", 0))
                if target in tasks:
                    tasks[target][1].cancelled.set()
                    tasks[target][0].cancel()
                response.update(type="beam.cancelled", payload={"ok": True})
            elif frame["type"] == "worker.shutdown":
                for task, beam in tasks.values():
                    beam.cancelled.set()
                    task.cancel()
                response.update(type="worker.stopped", payload={"ok": True})
                write_worker_frame(response)
                return 0
            else:
                raise RuntimeError(f"unknown frame type: {frame['type']}")
        except asyncio.CancelledError:
            failure_type = {
                "lens.view.request": "lens.view.result",
                "lens.derive.request": "lens.derive.result",
                "lens.coordinate.request": "lens.coordinate.result",
                "lens.query.request": "lens.query.result",
            }.get(frame["type"], "lens.refract.result")
            response.update(type=failure_type, payload={
                "ok": False, "error": {"code": "cancelled", "message": "beam cancelled"},
            })
        except Exception as exception:  # boundary conversion is intentional
            failure_type = {
                "lens.view.request": "lens.view.result",
                "lens.derive.request": "lens.derive.result",
                "lens.coordinate.request": "lens.coordinate.result",
                "lens.query.request": "lens.query.result",
            }.get(frame["type"], "lens.refract.result")
            response.update(type=failure_type, payload={
                "ok": False,
                "error": {"code": "lens_crashed", "message": str(exception), "retryable": False},
            })
        write_worker_frame(response)


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
